### PR TITLE
Toponaming/Part:  Bring over attacher differences

### PR DESCRIPTION
**Pull request by bgbsww**

### CODE REVIEW

File: src/Mod/Part/App/Attacher.cpp [hunk 1]
```diff
@@ -341,22 +397,23 @@ void AttachEngine::suggestMapModes(SuggestResult &result) const
         if (! this->modeEnabled[iMode])
             continue;
         const refTypeStringList &listStrings = modeRefTypes[iMode];
-        for (const auto & str : listStrings) {
+        for (std::size_t iStr = 0; iStr < listStrings.size(); ++iStr) {
             int score = 1; //-1 = topo incompatible, 0 = topo compatible, geom incompatible; 1+ = compatible (the higher - the more specific is the mode for the support)
+            const refTypeString &str = listStrings[iStr];
```
Comment (chennes) [inline src/Mod/Part/App/Attacher.cpp:402]: Why this change?

File: src/Mod/Part/App/Attacher.cpp [hunk 2]
```diff
@@ -394,7 +451,7 @@ void AttachEngine::suggestMapModes(SuggestResult &result) const
                 }
             }
             if (score > 0){
-                if(mlist.empty())
+                if(mlist.size() == 0)
```
Comment (chennes) [inline src/Mod/Part/App/Attacher.cpp:454]: Why this change?

File: src/Mod/Part/App/Attacher.cpp [hunk 3]
```diff
@@ -752,43 +809,54 @@ GProp_GProps AttachEngine::getInertialPropsOfShape(const std::vector<const TopoD
...
-    const std::vector<App::DocumentObject*> &objs = references.getValues();
-    const std::vector<std::string> &sub = references.getSubValues();
     geofs.resize(objs.size());
     storage.reserve(objs.size());
     shapes.resize(objs.size());
     types.resize(objs.size());
     for (std::size_t i = 0; i < objs.size(); i++) {
-        if (!objs[i]->isDerivedFrom<App::GeoFeature>()) {
-            throw AttachEngineException("AttachEngine3D: link points to something that is not App::GeoFeature");
+        if (!objs[i]->getTypeId().isDerivedFrom(App::GeoFeature::getClassTypeId())) {
```
Comment (chennes) [inline src/Mod/Part/App/Attacher.cpp:824]: This is a reversion to an older style of isDerivedFrom

**Review by chennes — COMMENTED** [review]
A couple more comments: I'm finding these by looking at the git blame and looking for changes in the last two years (which in some cases have not yet been adopted into LinkStage3).

File: src/Mod/Part/App/Attacher.cpp [hunk 4]
```diff
@@ -1700,253 +2003,301 @@ AttachEngineLine *AttachEngineLine::copy() const
...
-            gp_Pnt p2 = distancer.PointOnShape2(1);
-            LineBasePoint = p1;
-            gp_Vec dist = gp_Vec(p1,p2);
-            if (dist.Magnitude() < Precision::Confusion())
-                throw Base::ValueError("AttachEngineLine::calculateAttachedPlacement: can't make proximity line, because shapes touch or intersect");
-            LineDir = gp_Dir(dist);
-        }break;
-        default:
-            throwWrongMode(mmode);
+                throwWrongMode(mmode);
```
Comment (chennes) [inline src/Mod/Part/App/Attacher.cpp:2284]: What has happened to the `mm1Intersection` case that @LemonBoy added last month?

Comment (bgbsww) [reply src/Mod/Part/App/Attacher.cpp:2284]: Git rebase hates me, and maybe github hates you?  I swore you made this comment before and I saw it on my first scan, but when I was fixing comments I couldn't find it after looking several times the next day.  In any case, a manual restoration puts it back.  Good catch.  <later>  Oh, I get it, rebase thinks it was an intentional deletion, not a case of code ported from an older version.  Duh!
